src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HPack/HPackEncoder.cs

```diff
@@ -2,214 +2,151 @@
+
 namespace System.Net.Http.HPack
 {
     internal static class HPackEncoder
     {
         // Things we should add:
-        // * Static table encoding
-        //          This is implemented now for pseudo-headers, but not for regular headers.
-        //          This should be based off some sort of "known headers/values" scheme, so we don't need
-        //          to re-lookup header names each time.
         // * Huffman encoding
         //      
         // Things we should consider adding:
-        // * Dynamic table encoding.
-        //          This would make the encoder stateful, which complicates things significantly.
-        //          Additionally, it's not clear exactly what strings we would add to the dynamic table
-        //          without some additional guidance from the user about this.
-        //          So for now, don't do dynamic encoding.
+        // * Dynamic table encoding:
+        //   This would make the encoder stateful, which complicates things significantly.
+        //   Additionally, it's not clear exactly what strings we would add to the dynamic table
+        //   without some additional guidance from the user about this.
+        //   So for now, don't do dynamic encoding.
 
-        public static bool EncodeIndexedField(int index, Span<byte> buffer, out int length)
+        public static bool EncodeIndexedField(int index, Span<byte> destination, out int bytesWritten)
         {
-            length = 0;
-
-            if (buffer.Length == 0)
-            {
-                return false;
-            }
-
-            buffer[0] = 0x80;   // Literal field
-            if (!IntegerEncoder.Encode(index, 7, buffer, out length))
+            if (destination.Length != 0)
             {
-                return false;
+                destination[0] = 0x80;   // Literal field
+                return IntegerEncoder.Encode(index, 7, destination, out bytesWritten);
             }
 
-            return true;
+            bytesWritten = 0;
+            return false;
         }
 
-        public static bool EncodeIndexedName(int index, string value, Span<byte> buffer, out int length)
+        public static bool EncodeIndexedName(int index, string value, Span<byte> destination, out int bytesWritten)
         {
-            int i = 0;
-            length = 0;
-
-            if (buffer.Length == 0)
+            if (destination.Length != 0)
             {
-                return false;
-            }
-
-            buffer[0] = 0x40;   // Literal name
-            if (!IntegerEncoder.Encode(index, 6, buffer, out int indexLength))
-            {
-                return false;
-            }
-
-            i = indexLength;
-            if (i == buffer.Length)
-            {
-                return false;
-            }
-
-            if (!EncodeString(value, buffer.Slice(i), out int nameLength, lowercase: false))
-            {
-                return false;
+                destination[0] = 0x40;   // Literal name
```

Comment:
I think this is wrong. Sorry, my fault here.

0x40 will cause the header to be added to the dynamic table. This won't actually cause any harm, but since we aren't ever going to refer to it again, it makes more sense to encode it as a non-indexed header. Which is done via 0x00 here and passing 4 instead of 6 to the IntegerEncoder below. See https://tools.ietf.org/html/rfc7541#section-6.2.2.

Reply:
Ok